pkg/neonvm/controllers/vmmigration_controller.go

```diff
@@ -172,6 +172,26 @@ func (r *VirtualMachineMigrationReconciler) Reconcile(ctx context.Context, req c
 		if err := ctrl.SetControllerReference(vm, migration, r.Scheme); err != nil {
 			return ctrl.Result{}, err
 		}
+		// First check if VM has the affinity with CPU architecture
+		// if it doesn't we need to set it explicitly
+		// to prevent migration going to the node with different architecture
+		sourceNode := &corev1.Node{}
+		//nolint:exhaustruct // corev1.Node objects is not namespaced
+		err := r.Get(ctx, types.NamespacedName{Name: vm.Status.Node}, sourceNode)
+		if err != nil {
+			log.Error(err, "Failed to get source node")
+			return ctrl.Result{}, err
+		}
+		if !hasArchitectureAffinity(vm, sourceNode) {
+
+			if err := addArchitectureAffinity(vm, sourceNode); err != nil {
+				log.Error(err, "Failed to add architecture affinity to VM", "sourceNodeStatus", sourceNode.Status)
+			}
+			if err = r.Update(ctx, vm); err != nil {
+				log.Error(err, "Failed to update node affinity of the source vm")
+				return ctrl.Result{RequeueAfter: time.Second}, err
+			}
```

Comment:
question: IIUC, does this mean that the VM will permanently get affinity for the node it starts with?

Is there a reason we can't just do it on the target pod for the migration? (I guess, potential race conditions?)

Reply:
I don't think there is a particular reason, I just feel it is a bit safer to make the guard on the very first step of the migration. 

Plus if we going to have multiarch clusters it is a bit unsafe to even have VM without architecture affinity, may be in later it should be even added during the VM creation, in the vmcontroller?